File: clients/raycast/src/api/connection.ts
import { Alert, confirmAlert } from "@raycast/api";
import { Tag } from "./tag";
import { getServerUrl } from "./config";

export interface MetadataFieldType {
  id: number;
  name: string;
}

export interface ConnectionMetadata {
  id: number;
  field_type: MetadataFieldType;
  value: string;
}

export interface Connection {
  id: number;
  name: string;
  photo?: string;
  feed_count: number;
  uri_count: number;
  unread_uri_count: number;
  metadata: ConnectionMetadata[];
  tags: Tag[];
}

export interface ConnectionDetail {
  id: number;
  name: string;
  photo?: string;
  metadata: ConnectionMetadata[];
}

export interface ConnectionsResponse {
  data: Connection[];
  page: number;
  per_page: number;
  total: number;
  total_pages: number;
}

// Known field types matching server enum
export const FIELD_TYPES: MetadataFieldType[] = [
  { id: 1, name: "Bluesky" },
  { id: 2, name: "Email" },
  { id: 3, name: "GitHub" },
  { id: 4, name: "LinkedIn" },
  { id: 5, name: "Mastodon" },
  { id: 6, name: "Website" },
  { id: 7, name: "X" },
  { id: 8, name: "Other" },
];

export function listUrl({ page, query }: { page: number; query?: string }) {
  const params = new URLSearchParams({ page: String(page), per_page: "20" });
  if (query) {
    params.append("query", query);
  }
  return `${getServerUrl()}/connections?${params.toString()}`;
}

export async function getConnection(id: number): Promise<ConnectionDetail> {
  const response = await fetch(`${getServerUrl()}/connections/${id}`);
  if (!response.ok) {
    throw new Error("Failed to fetch connection");
  }
  return response.json();
}

export async function updateConnection(id: number, name: string, photo?: string | null): Promise<ConnectionDetail> {
  const body: { name: string; photo?: string | null } = { name };
  if (photo !== undefined) {
    body.photo = photo;
  }
  const response = await fetch(`${getServerUrl()}/connections/${id}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to update connection");
  }
  return response.json();
}

export async function deleteConnection(connection: Connection) {
  if (
    await confirmAlert({
      title: "Delete Connection",
      message: `Are you sure you want to delete "${connection.name}"?`,
      primaryAction: {
        title: "Delete",
        style: Alert.ActionStyle.Destructive,
      },
    })
  ) {
    await fetch(`${getServerUrl()}/connections/${connection.id}`, {
      method: "DELETE",
    });
  }
}

export async function createMetadata(
  connectionId: number,
  fieldTypeId: number,
  value: string,
): Promise<ConnectionMetadata> {
  const response = await fetch(`${getServerUrl()}/connections/${connectionId}/metadata`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ field_type_id: fieldTypeId, value }),
  });
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to create metadata");
  }
  return response.json();
}

export async function updateMetadata(
  connectionId: number,
  metadataId: number,
  value: string,
): Promise<ConnectionMetadata> {
  const response = await fetch(`${getServerUrl()}/connections/${connectionId}/metadata/${metadataId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ value }),
  });
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to update metadata");
  }
  return response.json();
}

export async function deleteMetadata(connectionId: number, metadataId: number): Promise<void> {
  const response = await fetch(`${getServerUrl()}/connections/${connectionId}/metadata/${metadataId}`, {
    method: "DELETE",
  });
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to delete metadata");
  }
}

// Refresh metadata types
export interface RefreshMetadataFeed {
  url: string;
  title?: string;
  format: "rss" | "atom" | "json_feed";
}

export interface RefreshMetadataProfile {
  url: string;
  field_type: MetadataFieldType;
}

export interface RefreshMetadataPreview {
  connection_id: number;
  source_url: string;
  proposed_name?: string;
  proposed_photo?: string;
  proposed_feeds: RefreshMetadataFeed[];
  proposed_profiles: RefreshMetadataProfile[];
  current_name: string;
  current_photo?: string;
  current_metadata: ConnectionMetadata[];
}

export async function fetchRefreshMetadataPreview(connectionId: number): Promise<RefreshMetadataPreview> {
  const response = await fetch(`${getServerUrl()}/connections/${connectionId}/refresh-metadata`);
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to fetch refresh metadata preview");
  }
  return response.json();
}

File: server/lib/handlers/connection.ml
open Tapak
open Handler_utils.Syntax
open Ppx_yojson_conv_lib.Yojson_conv.Primitives

(* Store sw and env for HTTP requests - set by main *)
let sw_ref : Eio.Switch.t option ref = ref None
let env_ref : Eio_unix.Stdenv.base option ref = ref None

let set_context ~sw ~env =
  sw_ref := Some sw;
  env_ref := Some env

let get_context () =
  match (!sw_ref, !env_ref) with
  | Some sw, Some env -> (sw, env)
  | _ -> failwith "Handler context not initialized"

let list request (pagination : Pagination.Pagination.t) =
  let query = Handler_utils.query "query" request in
  let* paginated =
    Service.Connection.list_with_counts ~page:pagination.page
      ~per_page:pagination.per_page ?query ()
    |> Handler_utils.or_connection_error
  in
  Handler_utils.json_response
    (Model.Connection.paginated_with_counts_to_json paginated)

let get_connection _request id =
  let* connection = Service.Connection.get ~id |> Handler_utils.or_connection_error in
  Handler_utils.json_response (Model.Connection.to_json connection)

type create_request = { name : string; url : string option [@yojson.option] }
[@@deriving yojson]

let create request =
  let* { name; url } =
    Handler_utils.parse_json_body create_request_of_yojson request
    |> Handler_utils.or_bad_request
  in
  if String.trim name = "" then Handler_utils.bad_request "Name cannot be empty"
  else
    let* connection =
      Service.Connection.create ~name () |> Handler_utils.or_connection_error
    in
    let () =
      match url with
      | Some url when String.trim url <> "" ->
          let website_field_type_id = Model.Metadata_field_type.(id Website) in
          let _ =
            Service.Connection_metadata.create ~connection_id:(Model.Connection.id connection)
              ~field_type_id:website_field_type_id ~value:url
          in
          ()
      | _ -> ()
    in
    Handler_utils.json_response ~status:`Created (Model.Connection.to_json connection)

type update_request = {
  name : string;
  photo : string option; [@yojson.option]
}
[@@deriving yojson]

let update request id =
  let* { name; photo } =
    Handler_utils.parse_json_body update_request_of_yojson request
    |> Handler_utils.or_bad_request
  in
  if String.trim name = "" then Handler_utils.bad_request "Name cannot be empty"
  else
    let* connection =
      Service.Connection.update ~id ~name ~photo |> Handler_utils.or_connection_error
    in
    Handler_utils.json_response (Model.Connection.to_json connection)

let delete_connection _request id =
  let* () = Service.Connection.delete ~id |> Handler_utils.or_connection_error in
  Response.of_string ~body:"" `No_content

(* Find the Website metadata field for a connection *)
let find_website_url (connection : Model.Connection.t) =
  List.find_opt
    (fun m ->
      Model.Metadata_field_type.equal
        (Model.Connection_metadata.field_type m)
        Model.Metadata_field_type.Website)
    (Model.Connection.metadata connection)
  |> Option.map Model.Connection_metadata.value

(* Refresh metadata preview - fetches metadata from connection's website and returns proposed changes *)
let refresh_metadata_preview _request id =
  let* connection = Service.Connection.get ~id |> Handler_utils.or_connection_error in
  let* website_url =
    find_website_url connection
    |> Handler_utils.or_bad_request_opt
         "Connection has no Website metadata field. Add a website URL first."
  in
  let* valid_url =
    Handler_utils.validate_url website_url |> Handler_utils.or_bad_request
  in
  let sw, env = get_context () in
  let* contact =
    Metadata.Contact.fetch ~sw ~env valid_url |> Handler_utils.or_bad_request
  in
  (* Build the preview response showing current vs proposed values *)
  let current_metadata = Model.Connection.metadata connection in
  let opt_field name = function
    | Some v -> [ (name, `String v) ]
    | None -> []
  in
  let feed_to_json (f : Metadata.Contact.Feed.t) =
    `Assoc
      ([
         ("url", `String f.url);
         ( "format",
           `String
             (match f.format with
             | Metadata.Contact.Feed.Rss -> "rss"
             | Atom -> "atom"
             | Json_feed -> "json_feed") );
       ]
      @ opt_field "title" f.title)
  in
  let profile_to_json (p : Metadata.Contact.Classified_profile.t) =
    `Assoc
      [
        ("url", `String p.url);
        ("field_type", Model.Metadata_field_type.to_json_with_id p.field_type);
      ]
  in
  let current_metadata_json =
    List.map Model.Connection_metadata.to_json current_metadata
  in
  let response =
    `Assoc
      ([
         ("connection_id", `Int id);
         ("source_url", `String valid_url);
       ]
      @ opt_field "proposed_name" contact.name
      @ opt_field "proposed_photo" contact.photo
      @ [
          ("proposed_feeds", `List (List.map feed_to_json contact.feeds));
          ( "proposed_profiles",
            `List (List.map profile_to_json contact.social_profiles) );
          ("current_name", `String (Model.Connection.name connection));
        ]
      @ opt_field "current_photo" (Model.Connection.photo connection)
      @ [ ("current_metadata", `List current_metadata_json) ])
  in
  Handler_utils.json_response response

let routes () =
  let open Tapak.Router in
  [
    get (s "connections")
    |> extract Pagination.Pagination.pagination_extractor
    |> request |> into list;
    get (s "connections" / int) |> request |> into get_connection;
    get (s "connections" / int / s "refresh-metadata")
    |> request |> into refresh_metadata_preview;
    post (s "connections") |> request |> into create;
    put (s "connections" / int) |> request |> into update;
    delete (s "connections" / int) |> request |> into delete_connection;
  ]
